### PR TITLE
56 expansion of heredocs

### LINKED ISSUE #56: Expansion of heredocs
Heredocs need to be expanded 

- For delimiter just quotes get removed.
- If delimiter was unquoted everything expands, regardless of single or double quotes. No field splitting or quote removal happens
- If delimiter was quoted nothing gets expanded
- This needs adjustment in expander, parser, executor to differentiate between quoted and unquoted delim

### PR DESCRIPTION
**Pull request by gwolf-011235**
Implement expansion of env vars in heredocs:

- Expander 
If HEREDOC found: Search and remove quotes from delim. If quotes were removed set type of "<<" to **QUOTEDOC**
 
- Parser 
Create array of bools. 
Sets the corresponding bool to true if QUOTEDOC is encountered.

- Expander
New struct  **t_hdoc** to save various heredoc related variables such as name and fd.
`ft_init_heredoc` sets all these variables. 
`ft_read_heredoc` checks with bool if delim was quoted. If delim wasn't quoted it expands env vars it finds.
For this new `ft_expander_heredoc` is called: only checks for $ and does no tilde expand or quote removal.

- Tests
Testcases for expander_heredoc added  

Closes #56 and #46

### CODE REVIEW

File: inc/mod_executor.h [hunk 1]
```diff
@@ -93,4 +102,7 @@ t_err	ft_unset(char **argv, t_hashtable *env_tab);
 t_err	ft_env(t_hashtable *env_tab);
 t_err	ft_exit(char **argv, t_hashtable *env_tab);
 
+//include from mod_expand
+t_err	ft_expander_heredoc(char **str, t_hashtable *symtab);
+
 #endif
```
Comment (u413-284-si) [inline inc/mod_executor.h]: newline missing

File: src/executor_heredoc.c [hunk 2]
```diff
@@ -89,24 +89,26 @@ t_err	ft_create_heredoc(t_cmd *cmd, char *delim, int curr_delim,
...
+		if (heredoc->quoted == false)
+			err = ft_expander_heredoc(&buf, symtab);
+		if (err != SUCCESS)
 			break ;
-		write(fd, buf, ft_strlen(buf));
-		write(fd, "\n", 1); 
+		ft_putendl_fd(buf, heredoc->fd);
 		free(buf);
 		buf = NULL;
 	}
```
Comment (u413-284-si) [inline src/executor_heredoc.c:115]: return err, if ERR_MALLOC was encountered during expanding

File: src/expand_expander.c [hunk 3]
```diff
@@ -50,8 +50,8 @@ t_err	ft_expander_arg(t_track *input, t_hashtable *symtab, t_type type)
 			err = ft_rm_single_quote(input);
 		else if (input->str[input->pos] == '"')
 			err = ft_rm_double_quote(input);
-		else if (input->str[input->pos] == '~' && (input->pos == 0
-				|| (type == ASSIGN && ft_strchr(input->str, '=')
+		else if (input->str[input->pos] == '~'
+			&& (input->pos == 0 || (type == ASSIGN && ft_strchr(input->str, '=')
 					== input->str + input->pos -1)))
```
Comment (u413-284-si) [inline src/expand_expander.c]: space after operator

**Comment by u413-284-si** [comment]
Splendid work! Integrates smoothly into existing code base providing enhanced features to our heredoc behaviour. No major issues found.

File: inc/mod_executor.h (hunk 1 above)
Comment (gwolf-011235) [reply inc/mod_executor.h]: added

File: src/executor_heredoc.c (hunk 2 above)
Comment (gwolf-011235) [reply src/executor_heredoc.c:115]: Added. Exact error control flow will be handled separately

File: src/expand_expander.c (hunk 3 above)
Comment (gwolf-011235) [reply src/expand_expander.c]: added

**Pull request merged**